scripts/rag/llama_index_w_evals_and_qa.py

```diff
@@ -0,0 +1,498 @@
+                if name == "multistep":
+                    df = pd.DataFrame(
+                        data,
+                        columns=[
+                            "query",
+                            "response",
+                            "response_evaluation",
+                            "response_latency",
+                        ],
+                    )
+                    all_data[chunk_size][name][k] = df
+                else:
+                    df = pd.DataFrame(data, columns=columns)
+                logger.info("RUNNING EVALS")
+                time_start = time.time()
+                df = df_evals(
+                    df=df,
+                    model=eval_model,
+                    formatted_evals_column="retrieval_evals",
+                    template=template,
+                )
+                time_end = time.time()
+                eval_latency = time_end - time_start
+                logger.info(f"EVAL LATENCY: {eval_latency:.2f}")
+                # Calculate MRR/NDCG on top of Eval metrics
+                df = calculate_metrics(df, k, formatted_evals_column="retrieval_evals")
+                all_data[chunk_size][name][k] = df
+
+            tmp_save_dir = save_dir + "tmp_" + str(chunk_size) + "/"
+            # Save tmp plots
+            plot_graphs(all_data=all_data, save_dir=tmp_save_dir, show=False)
+            # Save tmp raw data
+            with open(tmp_save_dir + "data_all_data.pkl", "wb") as file:
+                pickle.dump(all_data, file)
+
+    return all_data
+
+
+# Running the main Phoenix Evals both Q&A and Retrieval
+def df_evals(
+    df: pd.DataFrame,
+    model: BaseEvalModel,
+    formatted_evals_column: str,
+    template: str,
+):
+    # Then use the function in a single call
+    df["context"] = df["retrieved_context_list"].apply(
+        lambda chunks: concatenate_and_truncate_chunks(chunks=chunks, model=model, token_buffer=700)
+    )
+
+    df = df.rename(columns={"query": "question", "response": "sampled_answer"})
+    # Q&A Eval: Did the LLM get the answer right? Checking the LLM
+    Q_and_A_classifications = llm_eval_binary(
+        dataframe=df,
+        template=template,
+        model=model,
+        rails=["correct", "incorrect"],
+    )
+    df["qa_evals"] = Q_and_A_classifications
+    # Retreival Eval: Did I have the relevant data to even answer the question?
+    # Checking retrieval system
+
+    df = df.rename(columns={"question": "query", "retrieved_context_list": "reference"})
+    # query_column_name needs to also adjust the template to uncomment the
+    # 2 fields in the function call below and delete the line above
+    df[formatted_evals_column] = run_relevance_eval(
+        dataframe=df,
+        model=model,
+        template=templates.RAG_RELEVANCY_PROMPT_TEMPLATE_STR,
+        rails=list(templates.RAG_RELEVANCY_PROMPT_RAILS_MAP.values()),
+        # query_column_name="question",
+        # document_column_name="retrieved_context_list",
+    )
```

Comment:
We should create an issue to solve this @mikeldking. The problem is that the 2 fields commented don't update the variable name inside the function when creating prompts from `PromptTemplate`, so it errors because the variable `question` is not on the template. Happy to discuss options

Reply:
Issue: https://github.com/Arize-ai/phoenix/issues/1522